### ISSUE TITLE
[QUERY] Torn write when using BlobClient's UploadAsync

### ISSUE THREAD

**Issue by luckerby** [Azure/azure-sdk-for-net#15863]
Is it possible to run into a situation where a piece of code writes using `BlobClient`'s `UploadAsync`, while a consumer using the `DownloadAsync` would end up with garbled text?

An example: say a writer uploads 20 MB to a block blob, and a reader starts downloading the blob before the writer completes. Will the result of the read be part of those 20 MB, or would the reader be blocked until the upload completes?

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team best able to assist.

**Comment by seanmcc-msft** [comment]
Hi,

The download request will return an empty stream.  The `BlobClient.Upload()` API is really calling [Put Block](https://docs.microsoft.com/en-us/rest/api/storageservices/put-block) and [Put Block List](https://docs.microsoft.com/en-us/rest/api/storageservices/put-block-list) under the coverts.  The blob content does not become "real" until Put Block List is called.

-Sean

**Comment by luckerby** [comment]
@seanmcc-msft thanks! In other words, there's no need to concern myself with a synchronization mechanism, eg locks.

**Comment by seanmcc-msft** [comment]
Hi @luckerby, not in this case.  You would need to be careful though if you ended up writing to the same blob from multiple threads.

**Comment by luckerby** [comment]
@seanmcc-msft For the multiple writers case, it's probably due to the fact that one of them will "win" when doing [Put Block List](https://docs.microsoft.com/en-us/rest/api/storageservices/put-block-list), essentially overriding what others did.

Is there some synchronization attributes that could be used in the types present in the Azure SDK, or would using standard locking mechanisms (eg `lock`) be the way to go, in this multiple-writers scenario?

**Comment by seanmcc-msft** [comment]
@luckeryby, that is correct.  I recommend avoiding having multiple writers write to the same blob at the same time if possible.  If you can't avoid this, I think using `lock` would be the best way to go.
